core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java

```diff
@@ -234,78 +227,83 @@ public void addInput(Page page)
         for (int channelIndex = 0; channelIndex < channels.size(); ++channelIndex) {
+            ChannelState state = channelsState.get(channelIndex);
             Block block = page.getBlock(channels.get(channelIndex).index);
-            TypedSet valueSet = valueSets[channelIndex];
-            for (int position = 0; position < block.getPositionCount(); ++position) {
-                valueSet.add(block, position);
+            switch (state) {
+                case SET:
+                    TypedSet valueSet = valueSets[channelIndex];
+                    for (int position = 0; position < block.getPositionCount(); ++position) {
+                        valueSet.add(block, position);
+                    }
+                    filterSizeInBytes += valueSet.getRetainedSizeInBytes();
+                    filterMaxDistinctValues = Math.max(filterMaxDistinctValues, valueSet.size());
+                    break;
+                case MIN_MAX:
+                    BlockPositionComparison comparison = minMaxComparisons[channelIndex];
+                    updateMinMaxValues(block, channelIndex, comparison);
+                    break;
+                case NONE:
+                    break;
             }
-            filterSizeInBytes += valueSet.getRetainedSizeInBytes();
-            filterMaxDistinctValues = Math.max(filterMaxDistinctValues, valueSet.size());
         }
-        if (filterMaxDistinctValues > maxDistinctValues || filterSizeInBytes > maxFilterSizeInBytes) {
-            // The whole filter (summed over all columns) exceeds maxFilterSizeInBytes or a column contains too many distinct values
-            handleTooLargePredicate();
+        while (filterSizeInBytes > maxFilterSizeInBytes || filterMaxDistinctValues > maxDistinctValues) {
+            boolean biggestRetainedSize = filterSizeInBytes > maxFilterSizeInBytes;
+            int biggestChannel = getBiggestChannel(biggestRetainedSize);
+            boolean convertToMinMax = channelsState.fallback(biggestChannel);
+            if (convertToMinMax) {
+                BlockPositionComparison comparison = minMaxComparisons[biggestChannel];
+                Block block = blockBuilders[biggestChannel].build();
+                updateMinMaxValues(block, biggestChannel, comparison);
+            }
+
+            // recalculate max counters
+            filterSizeInBytes = 0;
+            filterMaxDistinctValues = 0;
+            for (int channelIndex = 0; channelIndex < channels.size(); channelIndex++) {
+                if (channelsState.get(channelIndex) == ChannelState.SET) {
+                    filterSizeInBytes += valueSets[channelIndex].getRetainedSizeInBytes();
+                    filterMaxDistinctValues = Math.max(filterMaxDistinctValues, valueSets[channelIndex].size());
+                }
+            }
         }
     }
 
-    private void handleTooLargePredicate()
+    private int getBiggestChannel(boolean biggestRetainedSize)
```

Comment:
It seems overly complicated to implement fallback for biggest channel.
Can we instead make a per channel `ChannelFilter` class and let each channel individually fallback to min-max and then no collection based on the distinct values and row count seen by it ?

Reply:
The problem with this approach is that the we need to either add more config options to an already bloated config class or change the meaning of the existing ones.
The solution with getting rid of the biggest df is the least problematic one

Reply:
I would rather remove/rename/repurpose configs to get simpler code here than make this code even more complex than it already is.
Typically we expect users to use `enable-large-dynamic-filters`, the rest of the configs are a last resort to allow dynamic filtering for specific scenarios. I don't see those configs as as some immutable contract.

Reply:
@sopel39 Please share your opinion here

Reply:
> The problem with this approach is that the we need to either add more config options to an already bloated config class or change the meaning of the existing ones.

I think we can keep current properties, but apply them per channel.
Per channel limits seem to make more sense.

Let's take semi join as an example:
plan1
```
semijoin(l1 = r1)
  semijoin(l2 = r2)
```
vs plan2
```
semijoin(l1 = r1 and l2 = r2)
```
it doesn't make sense that `plan2` has smaller limits than plan1

Reply:
All right, I will rewrite this to be on a per-channel basis and ping you when its done. The code should indeed be simpler.